### ISSUE TITLE
FFT processing in parallel group on supernova causes loud distortion (macOS)

### ISSUE THREAD

**Issue by dyfer** [supercollider/supercollider#4289]
## Environment

* SuperCollider version: 3.10.3
* Operating system: macOS 10.13.6
* Other details (Qt version, audio driver, etc.):

## Steps to reproduce
This was previous reported as issue with PartConv, but actually seems to apply to FFT processing in general on my machine.

```supercollider
(
//CAREFUL! this will create LOUD DISTORTION
~numSynths = 2;

~syncBetweenSynths = true; //works fine
// ~syncBetweenSynths = false; //THIS WILL CREATE LOUD DISTORTION

Server.supernova;

s.waitForBoot {

	SynthDef(\pt_test, {arg playbuf = 0, irbuf = 0, amp = 0.5, out = 0;
		var sig, chain;
		sig = SinOsc.ar(300, 0, amp);
		chain = FFT(LocalBuf(2048), sig);
		sig = IFFT(chain); // inverse FFT
		Out.ar(out, sig * amp);
	}).add;

	~target = ParGroup();
	s.sync;

	~allSynths = ~numSynths.collect({|inc|
		var synth = Synth(\pt_test, [\amp, -26.dbamp, \out, inc], ~target);
		if(~syncBetweenSynths, {s.sync});
		synth;
	});
}
)

//free everything
(
~target.free;
currentEnvironment.clear;
)
```

## Expected vs. actual behavior

Actual behavior:
When multiple ~~PartConvs~~ synths with FFT are started in a parallel group at the same time, they output loud distortion. If they are started after waiting for the server to allocate resources for the previous Synth, or if it's a regular group, they work fine.

Expected behavior: 
Multiple ~~PartConvs~~ synths with FFT  start simultaneously in a ParGroup without issue.

## Discussion:
It seems to be an issue with the vDSP implementation, as indicated in the [source](https://github.com/supercollider/supercollider/blob/develop/common/SC_fftlib.cpp#L24). Building supernova with the internal `Green FFT` library fixes the problem (although that's arguably not a proper "fix").

**Comment by patrickdupuis** [comment]
I wasn't immediately able to reproduce (Ubuntu, 3.10.2-rc1). I'll try this again later today.

**Comment by dyfer** [comment]
Few more observations:

- it does behave the same for me on macOS with 3.10.0ish, 3.10.1, 3.10.2-rc1
- additionally, it occurs also when more than one PartConv is started within a parallel group, _regardless of timing_, if
  - using built-in input/output on the macbook pro
  - not specifying hardware buffer size
- I tested on Ubuntu, and the error indeed did not occur

To clarify the second point - it works (and breaks) as described in the PR, with the exception of using built-in in/out _and_ not specifying hardware buffer size, when it breaks regardless of the  `syncBetweenSynths` flag (outputs loud noise). Maybe there is some kind of weird interaction between memory allocation and the PortAudio backend?

**Comment by sonoro1234** [comment]
What are the async things that should happen before sync?
I suppose that SynthDef sending only?

**Comment by dyfer** [comment]
@sonoro1234 
Do you mean in the reproducer for the bug?
Then it think the following should happen before the sync:
- sending a synthdef, 
- allocating playbuf and partconv buffers
- starting pargroup

**Comment by sonoro1234** [comment]
But the sync triggered by `~syncBetweenSynths` only waits for sending the synthdef and nothing else as the other async tasks already have a sync afterwards.
It is strange that only synthdef sending causes this!

**Comment by dyfer** [comment]
@sonoro1234 I now tested with creating a synthdef beforehand, so it's just waiting for starting the synth process itself (PartConv needs to allocate some memory as far as I understand). The result is the same in my testing. Here's the code:

```supercollider
(
//CAREFUL! this will create LOUD DISTORTION
//modified example from PartConv helpfile
~fftsize = 2048;
~numSynths = 2;

~syncBetweenSynths = true; //works fine
// ~syncBetweenSynths = false; //THIS WILL CREATE LOUD DISTORTION

Server.supernova;
s.options.memSize_(2.pow(16)); //if there is not enough realtime memory allocated, PartConv will crash the server
s.options.hardwareBufferSize_(256);

s.waitForBoot {
	var ir, irbuffer, bufsize;
	ir = [1] ++ 0.dup(100) ++ (
		(1, 0.99998 .. 0)
		.collect {|f|
			f = f.squared.squared;
			f = if(f.coin) { 0 }{ f.squared };
			f = if(0.5.coin) { 0 - f } { f }
		} * 0.1
	);
	ir = ir.normalizeSum;
	irbuffer = Buffer.loadCollection(s, ir);
	s.sync;
	
	SynthDef(\pt_test, {arg playbuf = 0, irbuf = 0, amp = 0.5, out = 0;
		var sig;
		var input = PlayBuf.ar(1, playbuf, loop:1);
		sig = PartConv.ar(input, ~fftsize, irbuf, 0.5) * -6.dbamp;
		Out.ar(out, sig * amp);
	}).add;

	bufsize = PartConv.calcBufSize(~fftsize, irbuffer);
	~irspectrum = Buffer.alloc(s, bufsize, 1);
	~irspectrum.preparePartConv(irbuffer, ~fftsize);
	s.sync;

	irbuffer.free; // don't need time domain data anymore, just needed spectral version
	~buffer = Buffer.read(s, Platform.resourceDir +/+ "sounds/a11wlk01.wav");
	~target = ParGroup();
	s.sync;

	~allSynths = ~numSynths.collect({|inc|
		var synth = Synth(\pt_test, [\playbuf, ~buffer, \irbuf, ~irspectrum, \amp, -26.dbamp, \out, inc], ~target);
		if(~syncBetweenSynths, {s.sync});
		synth;
	});
}
)

//free everything
(
~target.free;
~buffer.free;
~irspectrum.free;
currentEnvironment.clear;
)
```

**Comment by sonoro1234** [comment]
I could not find sync related actions on PartConv. There is no LOCK_SNDBUF_SHARED which could create problems between threads.
I remember that I did a correction on PartConv https://github.com/supercollider/supercollider/pull/2015
It is a complicated UGen and I did a rework for not needing the preparation phase which I am using in my personal work with supernova and dont show any problem. It is here https://github.com/sonoro1234/DWGReverb/blob/master/PartitionedConvolutionTrig.cpp, snappiz did the sclang part. Perhaps you could give it a try?
The whole repo is https://github.com/sonoro1234/DWGReverb

**Comment by dyfer** [comment]
@sonoro1234 I made some tests and it looks like FFT processing in general is affected, not only PartConv.

Could you confirm whether this happens on your machine? (the code in the first post is updated accordingly)

> by the way, I could not find the meaning of (1, 0.99998 .. 0) What is it doing?

This is another way of creating arithmetic series. Syntax is a little different than `Array.series` - it is `(first, second .. last)`.
```SuperCollider
(1, 0.99998 .. 0);
-> [ 1.0, 0.99998, 0.99996, 0.99994, 0.99992, 0.9999, 0.99988, 0.99986, 0.99984, 0.99982, 0.9998, 0.99978, 0.99976, 0.99974, 0.99972, 0.9997, 0.99968, 0.99966, 0.99964, 0.99962, 0.9996, 0.99958, 0.99956, 0.99954, 0.99952, 0.9995, 0.99948, 0.99946, 0.99944, 0.99942, 0.9994, 0.99938, 0.99936, 0.99934, 0.99932, 0.9993, 0.99928, 0.99926, 0.99924, 0.99922, 0.9992, 0.99918, 0.99916, 0.99914, 0.99912, 0.9991, 0.99908, 0.99906, 0.99904, 0.99902, 0.999, 0.99898, 0.99896, 0.99894, 0.99892, 0.9989, 0.99888, 0.99886, 0.99...etc...

(1, 0.99998 .. 0).size;
-> 50000
```

I haven't had a chance to test your convolver, sorry.

**Comment by sonoro1234** [comment]
SC win32 mingw compiled:
I have done the test. There is a variation in volume in each execution (with or without  ~syncBetweenSynths) caused by f.coin and 0.5.coin that makes difficult to make conclusions.
It seems to be a variation in stereo between  ~syncBetweenSynths false and true
The test with scsynth performs similar to supernova.
I still dont understand what sync is doing there as there is no async action involved.

**Comment by dyfer** [comment]
@sonoro1234 I'm very sorry! I The top example was not updated. I updated it now. (I must've been in the preview mode previously...)

The issue seems to be with FFT processing in general, not only PartConv.

**Comment by sonoro1234** [comment]
~syncBetweenSynths makes no difference for me on SC win32 mingw compiled with the updated example

**Comment by dyfer** [comment]
Further investigation - this might be a lead (see [source](https://github.com/supercollider/supercollider/blob/develop/common/SC_fftlib.cpp#L23)):
> vDSP uses a "SplitBuf" as an intermediate representation of the data.
For speed we keep this global, although this makes the code non-thread-safe.

Is the global `splitBuf` the culprit then? How would we address this maintaining performance?

**Comment by joshpar** [comment]
For supernova, do we know if vDSP is used? I see this kind of thing:

https://github.com/supercollider/supercollider/blob/develop/common/SC_fftlib.cpp#L360 <https://github.com/supercollider/supercollider/blob/develop/common/SC_fftlib.cpp#L360>

In the code that makes me wonder if the supernova versions are hitting vDSP code.

**Comment by sonoro1234** [comment]
I think dyfer means that this is a clue about multithreading issues (not that we use vDSP)
For fftw this is the doc
http://www.fftw.org/fftw3_doc/Thread-safety.html

**Comment by dyfer** [comment]
Supernova [hits SplitBuf creation when starting](https://github.com/supercollider/supercollider/blob/develop/common/SC_fftlib.cpp#L198), so I guess it does use vDSP?

**Comment by dyfer** [comment]
How would I test if this is the problem? Would I put locks around access to SplitBuf? (that's rather above my head... but I can tinker)

**Comment by sonoro1234** [comment]
> Supernova [hits SplitBuf creation when starting](https://github.com/supercollider/supercollider/blob/develop/common/SC_fftlib.cpp#L198), so I guess it does use vDSP?

https://github.com/supercollider/supercollider/blob/develop/common/SC_fftlib.cpp#L187

implies this is not executed

**Comment by dyfer** [comment]
> https://github.com/supercollider/supercollider/blob/develop/common/SC_fftlib.cpp#L187
> 
> implies this is not executed

I thought `SC_FFT_VDSP` it active on macOS, while `SC_FFT_GREEN` isn't, right?

When I set the breakpoint in line 198, it halted supernova on boot.

**Comment by dyfer** [comment]
To confirm: I forced building with

**Comment by dyfer** [comment]
To confirm: I forced building without vDSP, but with the internal `FFT_GREEN` library and the problem **did go away**. 

(In order to do that, I had to temporarily comment out few lines in server's cmake file starting [here](https://github.com/supercollider/supercollider/blob/develop/server/CMakeLists.txt#L18))

IMO this, combined with inability to reproduce this on linux and windows, does indicate that the problem is with how we use vDSP, likely with the global `SplitBuf` mentioned earlier. I'm all ears for what possible solutions to this would be...

**Comment by joshpar** [comment]
@dyfer - the issues you were having are still only with supernova, yes? Or did you hear issues in scsynth as well?

**Comment by dyfer** [comment]
> I wasn`t aware that macosx used vDSP.
> Why not allow macosx to work with FFTW?

I don't know. vDSP is what was used before. Also, as it's provided by Apple might be optimized better? Maybe I'll try running some tests...

**Comment by joshpar** [comment]
Green, then FFTW then vDSP was used on OS X. yes - vDSP provided performance improvements and quality ones as well. We should still use this in scsynth unless you uncover any issues (in my opinion).

**Comment by dyfer** [comment]
> @dyfer - the issues you were having are still only with supernova, yes? Or did you hear issues in scsynth as well?

Yes, as the title and tags indicate, this is only with supernova. However, any possible future multithreaded implementation is likely to have this problem if it uses vDSP, as indicated in the [comment in the source](https://github.com/supercollider/supercollider/blob/develop/common/SC_fftlib.cpp#L24).

**Comment by dyfer** [comment]
> Green, then FFTW then vDSP was used on OS X. yes - vDSP provided performance improvements and quality ones as well. We should still use this in scsynth unless you uncover any issues (in my opinion).

Are you suggesting using FFTW with SuperNova is a viable fix, as opposed to fixing vDSP?

**Comment by joshpar** [comment]
I think FFTW could be. That's the one I would try first.

**Comment by dyfer** [comment]
> I think FFTW could be. That's the one I would try first.

As far as I can see, using FFTW for supernova and vDSP for scsynth would not be straightforward, as we globally set which FFT library to use. Do you think this is still worth investigating?

**Comment by joshpar** [comment]
Others can chime in, but I would rather not lose vDSP on scsynth, and I am not sure what would be needed to fix vDSP on supernova.

**Comment by dyfer** [comment]
> Others can chime in, but I would rather not lose vDSP on scsynth, and I am not sure what would be needed to fix vDSP on supernova.

I'm absolutely not suggesting losing vDSP! However I think it would be worthwhile to make it multithread-friendly...

**Comment by sonoro1234** [comment]
CMakeLists modification would be the way to achieve vDSP for scsynth and FFTW for supernova.
But in case FFTW performance were better than vDSP I would not do that, just for simplicity!

**Comment by dyfer** [comment]
All right. I am able to build supernova with FFTW and scsynth with vDSP on macOS. However I'm having some build verification/linking problems...

FFTW with supernova indeed does "fix" this problem for now, so after figuring out the linking issue I'll submit a PR suggesting this as a solution.

**Comment by brianlheim** [comment]
From what I understand reading this, supernova uses vDSP in a non-threadsafe way. Can anyone clearly point to vDSP documentation and supernova code sections that indicate that? That would be very helpful for whoever wants to investigate further.

**Comment by brianlheim** [comment]
In the interest of keeping this issue focused, further discussion about which FFT library to use in which executable/OS should really take place on another ticket.

**Comment by dyfer** [comment]
> From what I understand reading this, supernova uses vDSP in a non-threadsafe way. Can anyone clearly point to vDSP documentation and supernova code sections that indicate that? That would be very helpful for whoever wants to investigate further.

AFAIU this doesn't have to do with how supernova uses vDSP, but with how our vDSP implementation is not thread-safe. This is [noted in the code](https://github.com/supercollider/supercollider/blob/develop/common/SC_fftlib.cpp#L24). Various functions in that file use the global `splitBuf` memory which makes it not thread-safe, AFAIU. Please let me know if I should provide more information - although this is the extent of my understanding here...

> In the interest of keeping this issue focused, further discussion about which FFT library to use in which executable/OS should really take place on another ticket.

Sure. I'm hoping to provide a PR with FFT library options and further conversation will happen there.

**Comment by dyfer** [comment]
Fixed in #4583
